### DISCUSSION TITLE
QT in GTK based desktop environment - Font sizes

### DISCUSSION THREAD

**Discussion by advice2020** [doublecmd/doublecmd#1459] [General]
Double Commander 1.2.0 alpha Qt5 (revision 11186) portable
Debian based Linux distro / GTK based DE

I was wondering if there is anyone currently using the QT5 based version of DC in a GTK based DE that has some knowledge about QT5 font size capabilities.
I am looking to be able to change specific fonts within specific QT based applications.  For example say you wanted to just increase the font size of the “menu bar” or just the font size of “tabs”.

Both GTK 2 & 3 offer this capability, but I have not been able to find a way to achieve this in QT5.  The only things I have seen are environment variables for things like “SCALE” & “DPI”.  These two EVs sound more like they are for changing the font / size of entire GUI which I am not looking to do.

GTK3 offers this in the gtk.css file
GTK2 offers the use of .gtkrc2.0 file or custom gtkrc file with environment variables.
(Which I received help to learn about them both from very kind people)
However as mentioned I have not been able to find anything like this in QT5.
QT is supposed to be a very customizable based toolkit, it seems very strange to me that something as simple as font sizes does not exist?

I am currently trying to make a decision on which version of DC I should use GTK2 or QT5, inside a GTK based DE.
If anyone has any information they could share about this I would really appreciate it.  Even if the answer is that it cannot be done, this would be disapointing but helpful to know.

Thank you to anyone who reads this and for any information provided

**Comment by j2969719** [comment]
qt5ct or qt6ct also allows to set custom css, customize base fonts, etc.

**Comment by advice2020** [comment]
Hey, good to see that you are able to access Github, thank you for your response.
Ahh I was not aware of this, when I saw "Style Sheets" tab in qt5ct I just overlooked this because I saw the QSS extensions shown in the list and didn't think it had anything to do with what I was looking for.  I didn't realize that QT decided to make their own version CSS, called QSS just to make things a little more complicated :)

I have seen that qt5ct has a "Fonts" tab but this is I believe a global setting that applies to all the fonts within a QT application.  I am going to post this info on the forum as well, but do you by any chance know if this QSS capability within qt5ct allows for similar functionality to the gtk.css CSS capabilities in GTK3?  Where users can change font sizes of specific items within specific applications?
For Example:
`.thunar menubar menuitem label {font-size: 16pt}`
allows the font size for just the menu bar within an application such as Thunar to change

I have looked this up but have only seen it mentioned that it can save all its settings into a QSS files.
Tricky because every time I try to look up how to do something on the user end for toolkit functionality, I always get a ton of information about coding, and not much how general users can apply this information.
I see that the qt5ct folder contains an empty "qss" folder for custom QSS config files, I will try to save a QSS file there, just didn't know if the "environment variable" (not sure if the right term for CSS / QSS) would work the same.
Was just wondering if you or anyone else knew if this was possible before spending time to figure out that it cannot.
Even if you don't though, thank you for your response, if this QSS works the way I am hoping, this could be great.

**Comment by j2969719** [comment]
>specific applications

https://wiki.archlinux.org/title/Qt#Qt_Style_Sheets

e.g.

```
QTabBar {
  font-size: 20pt;
}
```

`doublecmd -stylesheet=test.qss`

**Comment by advice2020** [comment]
@j2969719
I have not been able to get everything working perfectly yet, there are a lot of links within links to read through, but just wanted make this post because was excited about this.  I can't believe you were able to find this / figure this out, thank you so much for this, I didn't think this was going to be possible.

At the moment I am not sure where to put the "test.qss" file so that the DC command knows where to find it.
I am using a generated standalone version of DC 1.2.0 alpha QT so my command includes a path.
I tried 
`"$HOME/path-to-standalone-DC/doublecmd -stylesheet=test.qss"`
The ArchWiki page you mentioned showed the following as the command template 
`qt_application -stylesheet style.qss`
so I tried removing the equal sign but still didn't work
`"$HOME/path-to-standalone-DC/doublecmd -stylesheet test.qss"`

However, the reason why I am making this response is because, I did end up putting the "test.qs" file into the ".config/qt5ct/QSS" folder I mentioned.
I then opened qt5ct and went into the "Style Sheets" tab and checked the box for this file because shows up in the list.  When I do this, then all QT applications tabs font changes to this, which is great news because it shows that this can done, just wanted to share this because was excited to see this.
The testing I did may have been a little out of order, so I will try the doublecmd command again and see if it will be able to find the "test.qss" file in the qt5ct/qss folder mentioned,

Thank You again

@Anyone
Does anyone know if QSS files can be accessed without the need for qt5ct program in a GTK based DE?
The qt5ct program causes some graphical issues for me in DC with things like dialog boxes.  I was wondering if I would be able to continue to use "qt5-gtk2-platformtheme" instead and somehow be able to access a QSS file in a different way, if possible?

**Reply by j2969719** [reply]
`"$HOME/path-to-standalone-DC/doublecmd" -stylesheet="$HOME/mycooldir/mycool.qss"`

**Reply by JonnyTech** [reply]
Maybe try double dash before the option: `/path_to_DC/doublecmd --stylesheet=./mycool.qss`

**Comment by advice2020** [comment]
@JonnyTech 
Thank you for your response but I was not able to get this code to work on my machine.
Edit:
I apologize this does work if adding to the doublecmd.sh file I mention more about this below.

@j2969719 
Thank you for your response
I tried this out and but it says that it cannot find "mycooldir"
No I am just kidding :)
I tried this out and it works!
'"$HOME/path-to-standalone-DC/doublecmd" -stylesheet="$HOME/mycooldir/mycool.qss"'
I cannot believe you were able to figure this out, thank you very much for this.

I thought that I would have to use this with qt5ct but for anyone interested in this in the future, you don't.
I also tried with "qt5-gtk2-platformtheme" and the -stylesheet variable still works.
I just have to go through the list and narrow down what environment variables are linked to what, but as mentioned the QTabBar one works for tab fonts.  If I can figure out the other ones for menu bar, dialogs, etc. I will post them here in case anyone is ever interested in achieving this in the future.

Thank you again for your help.

@j2969719 or anyone else that maybe knows the answer
Solve one problem, and another one appears :)
I mentioned this on the other site, but the only issue now is that when you use a QSS file with qt5ct OR qt5-gtk-platform theme, this then causes issues with DC dialog boxes such as New directory, delete confirmation, Lua input field dialogs, etc.  The fields and buttons get chopped off on the right side of the dialogs.  If you move the dialog then the dialog appears to get refreshed and then looks fine.  
This only occurs when QSS file is linked, works fine when it is not.
If anyone has any ideas of why this is and how to fix it, please let me know.
Thank You

**Comment by j2969719** [comment]
`./mycool.qss`  should also work if qss is located in the directory from which doublecmd is launched e.g. `cd "$HOME/mydcdir/" && ./doublecmd -stylesheet=./mycool.qss`

out of curiosity tried the qss above and dont see any problems with the dialogs (archlinux qt6.6.1/qt5.15.12)

archwiki has a link to a list of supported things https://doc.qt.io/qt-5/stylesheet-reference.html

**Comment by advice2020** [comment]
@j2969719 
Thank you for your response.
I was able to get the other code to work that JonnyTech mentioned, I was trying this in a launcher, but when I added to the doublecmd.sh file it worked

Thank you for testing out the QSS file and the dialogs, that is strange, I am not sure why I am getting this?  I think I might have an idea of something to try, but if not then I am out of ideas to test.

Oh yeah I saw that link in the Arch link you originally mentioned, that is what I was referring to,  I have to go through that list to narrow down which EVs are what.
Skif_off helped me figure out how to achieve this with GTK2 EVs, if QT5 ones are similar,
some DC items use many of them (for example the Options window)
also some fonts for example, linked to dialog boxes also get applied within Option as well because they are both using the same EV so it can get tricky.
If I can narrow them down I will try to post results in case anyone is interested in the future.

Thank You again for figuring out the font size part though, so far really works well.

**Comment by advice2020** [comment]
Is there anyone using a Debian based GTK DE with DC QT5 version that would be willing to test out this issue I am experiencing caused by the QSS style sheet?
If anyone has any ideas what could be causing this please let me know?

Thank you for any help provided.

I downloaded Dolphin file manager to try testing this out with a different QT based program.  I linked the QSS style sheet and did NOT experience any issues dialog boxes.
I tried the delete confirmation & the new folder dialog boxes, both looked fine.
So it appears to only be happening with DC based dialogs.
Has anyone ever experienced something like this in any other QT programs?

Just hoping to get more confirmation before I report issue, if possible.
Thank you again to anyone who reads this

**Comment by advice2020** [comment]
Trying not to go too far with QT because if I cannot solve all these problems then not going to be able to use, but could still be helpful with other QT programs and for other DC QT users.

@j2969719
Would you by any chance know if anything could be added to the QSS file to mimic the global font setting offered in "qt5ct" fonts tab?
I am trying to just use the "qt5-gtk2-platformtheme" because I don't think I will really use any of the other features offered by qt5ct.
Just wanted to maybe be able to combine this global font setting along with specific EV items fonts.
Also I mentioned on the forum as well, QT6? is this only available in Arch?
Was wondering if I could maybe be able to skip over QT5 and go right to QT6 in hopes that all my QT5 issues would be solved.  Do you know if there is a DC QT6 available for Debian based DE?

Thank you again for all your help

**Comment by j2969719** [comment]
no idea, idont even use qss, default settings work well for me

yep, on arch the package is even available from the distro repository itself https://archlinux.org/packages/extra/x86_64/doublecmd-qt6/

**Comment by JonnyTech** [comment]
Works for me...

`$ doublecmd`

![image](https://github.com/doublecmd/doublecmd/assets/3584732/0db924f4-b6f3-4650-81b2-e7cd5a4ce1e2)

`$ doublecmd -stylesheet test.qss`

![image](https://github.com/doublecmd/doublecmd/assets/3584732/9694968d-67bf-48fb-94ce-b8011265a45b)

```
$ cat test.qss 
QTabBar {
  font-size: 20pt;
}
```

```
Double Commander
Version: 1.2.0 alpha
Revision: 11193
Commit: 6bbfb62
Build date: 2024/01/14
Lazarus: 2.2.6.0
FPC: 3.2.2
Platform: x86_64-Linux-qt5
OS version: MX 23.2 Libretto

Widgetset library: Qt 5.15.8, libQt5Pas 5.6.1
```

**Comment by advice2020** [comment]
@j2969719 
Ah okay, I thought you might know the answer.  That list of EVs just focuses on specific items (which is what I was looking for overall) I just figured if I could find a global setting as well I might be able to combine them.  I will see if I could find out how to do this global font setting.
Ahh you Arch users and your instant access to things! :)
I really want to try, on the forum it was said that it could be compiled, but I think a little over my head how to achieve this at the moment.
Anyways, thank you again for all your help, this -stylesheet option really solved what I was looking to achieve.

Although it created a new issue :)
A different Debian based DE user confirmed my dialog box issue with QSS.
No solution yet but at least I know that it is not just me.
Thanks again

**Reply by j2969719** [reply]
https://doublecmd.h1n.ru/viewtopic.php?p=42605#p42605

**Comment by advice2020** [comment]
@JonnyTech 
Thank you for testing this out, what you posted was only half the test, but did answer something I was wondering, because you also mentioned that you had slow launch times with QT5 version.
When you do this though with the QSS file example you have shown, would you be able to test out your dialog boxes and see if you notice issues on the right side where fields / text / buttons get chopped off.  
Dialog boxes such as delete confirmation, new directory (folder), etc.
Then try moving the dialog and notice that it refreshes and looks fine.
Thank You

As mentioned another user mentioned they were able to generate this as well, my guess based on the information that you provided in your last post is that you are going to experience this issue as well.
Thank you again for your help

**Reply by JonnyTech** [reply]
> would you be able to test out your dialog boxes and see if you notice issues on the right side where fields / text / buttons get chopped off

@advice2020 certainly, I can test when I get home later - with the same .qss file or different contents? Can you provide a screenshot so that I know if I am looking at the correct place?

**Comment by advice2020** [comment]
@JonnyTech 
Oh, thank you.
Same .qss file is fine (turns out this even occurs if the .qss file is blank), just has to be a .qss file linked to the command using -sheetstyle exactly how you did it when you generated the results you shared.
I like using the example with the large QTabBar tab font because it lets me know that the QSS file has been loaded correctly.

Doing exactly what you did the last time, then can try the following two dialogs.
1.) Hit F7 (or whatever you use) to create a new folder/directory.
The dialog will say "Create new directory" in the header and you will not be able to miss it, if you look at the right side of the dialog you will see that the input field is chopped off and you only see some of the "OK" button the "Cancel" button to its right is not visibile because chopped off.
2.) Can also create a blank text file or empty folder and hit the "delete" key, similar concept, the delete confirmation dialog will appear (as long as yours its set to show this) and the "OK" & "Cancel" buttons probably will not even be visible on right side either.

If you move either of these dialog boxes mentioned above with your mouse, you will notice that they refresh and then look fine.

This will not cause any issues with you machine, once you remove the -stylesheet from the command everything works again, it is something with linking QSS file.
I am almost certain you will experience the same issue I mentioned.
I eventually will have to try this in a different distro / desktop environment to see if this occurs.

Again if you don't have time to test, don't worry about it, but thank you as mentioned your last post confirmed something for me which was already helpful.
Thanks again

**Reply by JonnyTech** [reply]
1.

Normal
![image](https://github.com/doublecmd/doublecmd/assets/3584732/03251739-8bc6-471a-b0a7-e4490395c472)

Stylesheet
![image](https://github.com/doublecmd/doublecmd/assets/3584732/a6d9c0f7-1d7c-4ed5-8d9f-1488cd3a9415)

Dragged
![image](https://github.com/doublecmd/doublecmd/assets/3584732/fde96b2b-ef7e-4231-88ef-3c6398adbf7d)

2.

Normal
![image](https://github.com/doublecmd/doublecmd/assets/3584732/2b134733-147c-4785-827a-e1901b0d8b1c)

Stylesheet
![image](https://github.com/doublecmd/doublecmd/assets/3584732/d6740679-45ef-4050-b1f8-4b066d037dca)

Dragged
![image](https://github.com/doublecmd/doublecmd/assets/3584732/c5a53426-f0c8-4203-b67f-894d95f10c65)

Screenshots are worth many words ;)

**Comment by advice2020** [comment]
@JonnyTech 
Thank you very much for testing / confirming this for me and providing the visuals.
I was pretty sure that you would experience these same results.

I tested some other Distros (all Debian based) / DEs and this result seems to be common.
One DE was even stranger, if DC was maximized then the dialog box would be fine, but if the DC window was smaller then this dialog issue would occur.
I don't know enough about QT variables, but I wonder if there is some way to be able to add some kind of "refresh" action in that test.qss file to refresh a dialog boxes, because as you can see when you drag the window it does refresh and looks fine.

I know that it is unlikely that a distro has anything to do with it, but j2969719 using Arch appears to be the only person that does not experience these results.

Anyways, thank you again for your help

If anyone has any other ideas please let me know.

**Comment by j2969719** [comment]
also tried it with an unmaximized window and also could not reproduce it. looked at the version information from jonnytech and compiled doublecmd with lazarus 2.2.6 and then the problem finally appeared, whatever it was, it seems to have been fixed in v3.0. it seems alex's suspicions are correct.  

you can try downloading lazarus from https://www.lazarus-ide.org and installing v3.0 and building doublecmd (to create a portable version, just run `./create_package.sh --ws=qt5 -P` in the 'install' folder, if successful, the archive should appear in the 'linux/release' subfolder)

**Reply by JonnyTech** [reply]
> it seems to have been fixed in v3.0

Confirmed

**Comment by advice2020** [comment]
@j2969719 
Wow, this was incredibly helpful, thank you so much for taking the time to do this.
I was trying so many different tests to try to narrow this down, what a waste of time, wow.
As mentioned, all of this just to change a couple of font sizes :)
Ahhhh the main thing is that I at least have an answer for what was going on.

Quick question, so you compiled your QT5 & QT6 versions yourself, that is why you have Lazarus 3.0 and not the 2.2.6 found in the alpha versions?

Again, thank you so much for all your help with getting an answer for this, as well as for the help with initial purpose which was being able to change the font sizes, very appreciated.

@JonnyTech 
Thank you again for your testing and I just saw, thank you for confirming 3.0 as well.

--------------------------------------------------------------------------------------
Okay, have never really compiled anything like this before, so I have to think about if this is going to be practical to get into.
Just lack of experience with this, but recently hearing that people are compiling using QT6, newer versions of Lazarus, etc. as mentioned I would have thought that DC would have to be compiled with the specific versions of things like these that it was coded to work with.
Just don't understand how you would know what to include when compiling, if using newer versions of packages, would imagine certain things would change.
So must just be for stability reasons that newer DC alpha versions don't just include newer versions of things such as Lazarus?
I like to be able to download the new alpha versions every now and then to keep my testing of the program up to date.
I don't know how difficult it would be to have to keep on compiling just to be able to keep on updating Lazarus to correct this issue.

Anyways, thank you both again for all your help, I really appreciate it.

**Reply by j2969719** [reply]
well, even the doublecmd v1.1.9 package from the archlinux repo is also compiled with lazarus v3.0..
